Topic Discovery

dennis-adlink opened the pull request:
This PR adds the implementation of DDS topic discovery, to enable the exchange of topic meta-data between endpoints
using the built-in topic announcer and detector endpoints and allow applications to read topic discovery data. For implementation details see the commit messages. 

There are a few open issues:
- Secure equivalents of the topic discovery (and type discovery?) endpoints may be required
- Currently only new topic definitions are written as DCPSTopic samples (which do not include the actual topic guid). Another built-in local writer could be added that writes updates for each discovered instance of these topic definitions (i.e. write a sample for each received SEDP topic sample). 

Because this implementation is building on type discovery (PR #559), the branch for this PR is branched from that code and the commits of that PR are included here. 

PS: commit 87cceee contains an implementation of a generic templated list, which is used at a few places in the type/topic discovery code. To simplify the rebasing mentioned above, I've decided to include that commit (which could have been a separate PR) in this PR.

k0ekk0ek commented on src/core/ddsc/src/dds_domain.c:
```diff
@@ -365,3 +365,79 @@ void dds_write_set_batch (bool enable)
+    rc = DDS_RETCODE_PRECONDITION_NOT_MET;
+    goto failed;
+  }
+  if (tlm->state == TL_META_RESOLVED)
+  {
+    *sertype = ddsi_sertype_ref (tlm->sertype);
+    ddsrt_mutex_unlock (&gv->tl_admin_lock);
+  }
+  else
+  {
+    ddsrt_mutex_unlock (&gv->tl_admin_lock);
+    if (!ddsi_tl_request_type (gv, &type_id))
+    {
+      rc = DDS_RETCODE_PRECONDITION_NOT_MET;
+      goto failed;
+    }
+
+    const dds_time_t tnow = dds_time ();
+    const dds_time_t abstimeout = (DDS_INFINITY - timeout <= tnow) ? DDS_NEVER : (tnow + timeout);
+    *sertype = NULL;
+    ddsrt_mutex_lock (&gv->tl_admin_lock);
+    // type may already be resolved at this point, which means we
+    // shouldn't wait for the condition to be triggered
+    if (tlm->state == TL_META_RESOLVED)
+      *sertype = ddsi_sertype_ref (tlm->sertype);
+    while (*sertype == NULL && dds_time () < abstimeout)
+    {
+      if (ddsrt_cond_waituntil (&gv->tl_resolved_cond, &gv->tl_admin_lock, abstimeout))
+      {
+        if (tlm->state == TL_META_RESOLVED)
+          *sertype = ddsi_sertype_ref (tlm->sertype);
+      }
+    }
+    ddsrt_mutex_unlock (&gv->tl_admin_lock);
+    if (*sertype == NULL)
+    {
+      rc = DDS_RETCODE_TIMEOUT;
+      goto failed;
+    }
+  }
```
I feel this is better rewritten(?) Basically you check for `TL_META_RESOLVED` twice, so it reads like these branches can be merged and perhaps of checking if `sertype` is NULL, just check `tlm->state` and make `sertype == NULL` an assert? Judging by the first branch that would be correct.

dennis-adlink replied:
You're right, this can be simplified with duplicate checks removed, I've fixed this in bf8af34

k0ekk0ek commented on src/core/ddsc/src/dds_whc_builtintopic.c:
```diff
@@ -86,49 +90,107 @@ static bool bwhc_sample_iter_borrow_next (struct whc_sample_iter *opaque_it, str
+        struct proxy_topic *proxytp = NULL;
+        if (it->cur_proxypp != NULL)
+        {
+          ddsrt_mutex_lock (&it->cur_proxypp->e.lock);
```
There's a little smell here. You take the lock conditionally, but unlock unconditionally. I think it would be cleaner to move this into the `while` loop to?

dennis-adlink replied:
There was an unnecessary `if` and an unreachable `return false` after the `while` loop, I've removed that in bf8af34. I think the conditions for lock and unlock are correct: the first lock, and unlock in case there is no next topic for the current proxypp, should be conditional because this should not be done in the first iteration. The last unlock can only be reached with `it->cur_proxypp->e.lock` taken, so there is no condition needed there. I've added some code comments in bf8af34 that hopefully make the code a bit more clear.

k0ekk0ek reviewed (commented):
Nice @dennis-adlink! I had a quick look, there's three things that stood out too me. I'm always a little concerned when locks are taken conditionally, so they might be entirely correct, but I think they can be simplified a little. I'm not going to (dis)approve, since you've been working with @eboasson on this, but that's what I noticed.

dennis-adlink commented on src/CMakeLists.txt:
```diff
@@ -79,6 +90,14 @@ if(ENABLE_DEADLINE_MISSED)
   add_definitions(-DDDSI_INCLUDE_DEADLINE_MISSED)
 endif()
 
+if(ENABLE_TYPE_DISCOVERY)
+  add_definitions(-DDDSI_INCLUDE_TYPE_DISCOVERY)
+endif()
+
+if(ENABLE_TOPIC_DISCOVERY)
+  add_definitions(-DDDSI_INCLUDE_TOPIC_DISCOVERY)
+endif()
+
```
I agree that it would be better to narrow the scope for these definitions, but downside is that the definitions cannot be used outside ddsc/ddsi. Currently the `ddsls` tool also uses it for including (or not) DCPSTopic support, so we'll also need to define it for that target (which cannot be done in /src/CMakeLists.txt, because the targets are not yet created at that point).

dennis-adlink commented:
Force pushed this branch because of a rebase on the type discovery branch (PR #559).

eboasson commented on src/core/ddsi/src/q_entity.c:
```diff
@@ -5414,33 +5625,51 @@ static void delete_ppt (struct proxy_participant *proxypp, ddsrt_wctime_t timest
-  ddsrt_free (eps);
+  ddsrt_free (child_entities);
+
+#ifdef DDS_HAS_TOPIC_DISCOVERY
```
The intent here is to delete all proxy topics of this proxy participant, right? So why go through the complications of `delete_proxy_topic`, which locks the proxy participant and searches for the given definition? It's fortunate that the order of the searching is the same as the order of deleting them, otherwise it'd be O(n^2)!

I do see the locking order is a bit in the way, but it seems that it would be safe to do just delete them all earlier, where `deleting` is set. At first sight it seems that that would remove the need to build `topic_refs` as well.

eboasson commented on src/core/ddsi/src/q_entity.c:
```diff
@@ -5570,6 +5799,236 @@ uint64_t get_entity_instance_id (const struct ddsi_domaingv *gv, const struct dd
+  proxy_topic_list_iter_t it;
+  struct topic_definition *tpd = lookup_topic_definition (gv, qos, type_id, NULL, &new_tpd);
+  ddsrt_mutex_lock (&proxypp->e.lock);
+  for (struct proxy_topic *proxytp = proxy_topic_list_iter_first (&proxypp->topics, &it); proxytp != NULL && !found_proxytp; proxytp = proxy_topic_list_iter_next (&it))
```
I am not entirely happy with all the linear scans to find topics (by name in`lookup_topic_definition_by_name` , by definition here, in `delete_proxy_topic`, perhaps also elsewhere): before you know it all starts to add up to real work.

That said, I don't want to derail this PR by asking for a rewrite of all this. It is optional (and excluded by default) so we can accept a suboptimal implementation for the time being.

eboasson commented on src/core/ddsi/src/q_entity.c:
```diff
@@ -5914,17 +6376,21 @@ int delete_proxy_writer (struct ddsi_domaingv *gv, const struct ddsi_guid *guid,
   GVLOGDISC ("- deleting\n");
   builtintopic_write (gv->builtin_topic_interface, &pwr->e, timestamp, false);
+#ifdef DDS_HAS_TYPE_DISCOVERY
+  /* Unref tl_meta before removing from entity index, because
```
I think I get the reasoning; the bit I am not yet sure of is what the consequences of unnecessarily triggering an update for the endpoint matching would be. I think it would simply be a waste of CPU cycles. Am I right?

dennis-adlink replied:
Yes that's correct, and this will not cause an extra trigger of the endpoint matching. This only ensures that the proxy endpoint that's being deleted will not be registered with the type anymore, so that `ddsi_tl_handle_reply` will not try to do any matching for this proxy endpoint.

eboasson reviewed (commented):
Thanks for this work @dennis-adlink! And please accept apologies for the delay in reviewing it. The general comment is that I am fine with it, but there are some things that do require a bit of rework. The most important is probably the way the built-in topic WHC does the iteration of proxy topics, I believe there's a race condition between deleting a proxy participant and the iteration through its topics.

Another — unrelated — aspect is that if one publishes the topics in discovery, then one should also do updates to it when its QoS changes. The only change permitted by Cyclone is `TOPIC_DATA`, but still ...

There are some other things as well, some of them really minor. I'd appreciate it if you would have a look at my comments and either rework things or discuss them.

dennis-adlink commented on src/core/ddsi/include/dds/ddsi/q_rtps.h:
```diff
@@ -74,6 +74,8 @@ typedef int64_t seqno_t;
 #define NN_ENTITYID_KIND_WRITER_NO_KEY 0x03
 #define NN_ENTITYID_KIND_READER_NO_KEY 0x04
 #define NN_ENTITYID_KIND_READER_WITH_KEY 0x07
+#define NN_ENTITYID_KIND_CYCLONE_TOPIC_BUILTIN 0x0c // Vendor specific
```
When allocating a new entity ID (in `ddsi_new_topic`), the flag `NN_ENTITYID_SOURCE_VENDOR` is added so that the topic entities are marked as vendor specific. I think that the two kinds we add here should fit in the kind mask (0x3f), and I also think the kind 'topic' should actually have been defined in the RTPS specification. The reason for defining two topic kinds here (which is far from ideal), is because built-in and vendor-specific flag values cannot be combined (see RTPS 2.3 spec section 9.3.1.2)

dennis-adlink commented on src/core/ddsi/include/dds/ddsi/ddsi_entity_index.h:
```diff
@@ -111,6 +100,9 @@ struct entidx_enum_proxy_participant { struct entidx_enum st; };
 struct entidx_enum_proxy_writer { struct entidx_enum st; };
 struct entidx_enum_proxy_reader { struct entidx_enum st; };
 
+struct match_entities_range_key *entidx_minmax_new (void);
```
The reason I've added an allocation function is that `struct match_entities_range_key` now has a `struct topic`, which has a pointer to a `ddsi_topic_definition`. The qos in `match_entities_range_key` for a topic entity is in the topic definition (`min->entity.tp.definition->xqos`), so the topic definition needs to be allocated. To force using the allocation function, I've made the type opaque. For now I'll rename the `fini` function, but if you have an alternative for using allocation, then I'm fine with changing that as well.

dennis-adlink commented:
@eboasson thanks for your extensive review! I think commit 7a46a71 addresses all issues and suggestions from your review, see the commit message for an overview of what I've changed. For some of the comments I've written a reply above. I think that both topic QoS updating and proxy topic deletion requires some additional tests, so I'll add these later this week.

dennis-adlink commented on src/core/ddsc/src/dds_whc_builtintopic.c:
```diff
@@ -99,36 +104,96 @@ static bool bwhc_sample_iter_borrow_next (struct whc_sample_iter *opaque_it, str
+        if (it->cur_proxypp != NULL)
+        {
+          ddsrt_mutex_lock (&it->cur_proxypp->e.lock);
+          proxytp = proxy_topic_list_iter_next (&it->proxytp_it);
```
As discussed (offline), I have changed the deletion of proxy topics: the delete function marks the proxy topic as 'deleted' and the GC delete will remote it from the proxy topic list of the proxy participant. This guarantees a valid pointer when getting a proxy topic from the proxy participant's list, and we can just skip the ones that have 'deleted' set. It also 
eliminates the need to keep track of items removed from the proxy topics list during the iteration of this list in the whc.

dennis-adlink commented on src/core/ddsi/include/dds/ddsi/q_whc.h:
```diff
@@ -58,6 +58,7 @@ struct whc_sample_iter {
     double y;
     void *p;
   } opaque;
+  uint8_t pad; /* so that bwhc_iter fits in */
```
Yes, that `pad` is obviously wrong! I completely missed that line and the code comment.

dennis-adlink commented on src/core/ddsi/src/ddsi_security_omg.c:
```diff
@@ -1714,6 +1714,15 @@ unsigned determine_publication_writer (const struct writer *wr)
+#ifdef DDS_HAS_TOPIC_DISCOVERY
+unsigned determine_topic_writer (const struct topic *tp)
+{
+  // FIXME: secure topic writer required?
```
I've changed this so that the implementation will `abort` in case topic discovery is used (not only enabled the build option, but also the topic discovery endpoints enabled in config) and discovery protection is enabled in the participant.

dennis-adlink commented on src/core/ddsi/src/q_ddsi_discovery.c:
```diff
@@ -1400,12 +1534,24 @@ static void handle_SEDP (const struct receiver_state *rst, seqno_t seq, struct d
-        handle_SEDP_dead (rst, &decoded_data, serdata->timestamp);
+        if (!is_topic)
+          handle_SEDP_dead (rst, &decoded_data, serdata->timestamp);
+        /* for proxy_topic a SEDP dispose/unregister is ignored (and is not sent by Cyclone)
```
You're right, that comment was outdated, the SEDP dispose message are actually sent, and we need to handle these messages and remove the corresponding proxy topics. I've changed this in 7a46a71

dennis-adlink commented:
When I added the additional tests, I found a few other issues. In commit f4d9a20 I've added fixes for these bugs, as well as a some other refactoring (see the commit message). The QoS update test also revealed a memory leak in cleaning up a reader's loan in a scenario where the reader (or participant/domain) is deleted asynchronously while the application is reading data and returning the loan (which is of course not a very common scenario, but it should work anyway). I've fixed this in a separate PR #689.

eboasson commented:
@dennis-adlink I have spent a bit of time with it, rebased it to master (with a bit of squashing to keep it simple) and a few small changes:
* removal of dynamic memory allocation for entity index iterators (something discussed above in the comments)
* a bit of refactoring in iterating over the topics in the WHC for built-in topics
* always defining `dds_builtintopic_topic_t`, I think that works out better for writing application code that wants to utilise topic discovery but can live without
* fixed init/destroy of some state in case of type discovery but no topic discovery
* tweaked the set of CI builds to include a build with type discovery but without topic discovery (plus some related things)

all this is at https://github.com/eboasson/cyclonedds/tree/topic-discovery-pr615

So that's all fine, I could squash it and call it a day. However, I am running into a deadlock on Travis:
```
1615552516.580646 [1]  threadmon: thread 26840 failed to make progress
29331615552516.580659 [1]  threadmon: thread gc failed to make progress
29341615552516.580662 [1]  threadmon: thread dq.builtins failed to make progress
29351615552516.580665 [1]  threadmon: -- stack trace of 26840 requested --
29361615552516.581786 [1]  threadmon: -- stack trace follows --
29371615552516.586064 [1]  threadmon: 0   libddsc.0.7.0.dylib                 0x000000010e50effe log_stacktrace_sigh + 30
29381615552516.586074 [1]  threadmon: 1   libsystem_platform.dylib            0x00007fff69d515fd _sigtramp + 29
29391615552516.586077 [1]  threadmon: 2   libsystem_malloc.dylib              0x00007fff69d11297 tiny_malloc_should_clear + 288
29401615552516.586078 [1]  threadmon: 3   libsystem_pthread.dylib             0x00007fff69d58937 _pthread_mutex_firstfit_lock_slow + 222
29411615552516.586081 [1]  threadmon: 4   libddsc.0.7.0.dylib                 0x000000010e53d879 ddsrt_mutex_lock + 9
29421615552516.586083 [1]  threadmon: 5   libddsc.0.7.0.dylib                 0x000000010e5111d4 tlm_register_with_proxy_endpoints_locked + 308
29431615552516.586171 [1]  threadmon: 6   libddsc.0.7.0.dylib                 0x000000010e511063 ddsi_tl_meta_register_with_proxy_endpoints + 163
29441615552516.586174 [1]  threadmon: 7   libddsc.0.7.0.dylib                 0x000000010e5230cc dds_create_topic_impl + 2092
29451615552516.586176 [1]  threadmon: 8   libddsc.0.7.0.dylib                 0x000000010e5235bc dds_create_topic + 700
29461615552516.586177 [1]  threadmon: 9   cunit_ddsc                          0x000000010e450285 CU_TestProxy_ddsc_topic_discovery_remote_topics + 661
29471615552516.586181 [1]  threadmon: 10  libcunit.1.dylib                    0x000000010e49059f run_single_test + 143
29481615552516.586182 [1]  threadmon: 11  libcunit.1.dylib                    0x000000010e48ffc5 run_single_suite + 325
29491615552516.586184 [1]  threadmon: 12  libcunit.1.dylib                    0x000000010e48fdd8 CU_run_all_tests + 216
29501615552516.586186 [1]  threadmon: 13  cunit_ddsc                          0x000000010e3da4e8 main + 51832
29511615552516.586187 [1]  threadmon: 14  libdyld.dylib                       0x00007fff69b54cc9 start + 1
29521615552516.586189 [1]  threadmon: -- end of stack trace --
29531615552516.586191 [1]  threadmon: -- stack trace of gc requested --
29541615552516.587902 [1]  threadmon: -- stack trace follows --
29551615552516.587995 [1]  threadmon: 0   libddsc.0.7.0.dylib                 0x000000010e50effe log_stacktrace_sigh + 30
29561615552516.588003 [1]  threadmon: 1   libsystem_platform.dylib            0x00007fff69d515fd _sigtramp + 29
29571615552516.588005 [1]  threadmon: 2   ???                                 0x00007fb8c15315d3 0x0 + 140431494157779
29581615552516.588007 [1]  threadmon: 3   libsystem_pthread.dylib             0x00007fff69d58937 _pthread_mutex_firstfit_lock_slow + 222
29591615552516.588008 [1]  threadmon: 4   libddsc.0.7.0.dylib                 0x000000010e53d879 ddsrt_mutex_lock + 9
29601615552516.588010 [1]  threadmon: 5   libddsc.0.7.0.dylib                 0x000000010e50fe93 tlm_unref_impl + 611
29611615552516.588011 [1]  threadmon: 6   libddsc.0.7.0.dylib                 0x000000010e4ee9a9 gc_delete_proxy_topic + 169
29621615552516.588013 [1]  threadmon: 7   libddsc.0.7.0.dylib                 0x000000010e4f039d gcreq_queue_thread + 157
29631615552516.588014 [1]  threadmon: 8   libddsc.0.7.0.dylib                 0x000000010e5061d3 create_thread_wrapper + 147
29641615552516.588016 [1]  threadmon: 9   libddsc.0.7.0.dylib                 0x000000010e53e0c9 os_startRoutineWrapper + 25
29651615552516.588017 [1]  threadmon: 10  libsystem_pthread.dylib             0x00007fff69d5d109 _pthread_start + 148
29661615552516.588019 [1]  threadmon: 11  libsystem_pthread.dylib             0x00007fff69d58b8b thread_start + 15
29671615552516.588021 [1]  threadmon: -- end of stack trace --
29681615552516.588023 [1]  threadmon: -- stack trace of dq.builtins requested --
29691615552516.590247 [1]  threadmon: -- stack trace follows --
29701615552516.590334 [1]  threadmon: 0   libddsc.0.7.0.dylib                 0x000000010e50effe log_stacktrace_sigh + 30
29711615552516.590337 [1]  threadmon: 1   libsystem_platform.dylib            0x00007fff69d515fd _sigtramp + 29
29721615552516.590339 [1]  threadmon: 2   libddsc.0.7.0.dylib                 0x000000010e4e4919 writer_max_drop_seq + 25
29731615552516.590340 [1]  threadmon: 3   libsystem_pthread.dylib             0x00007fff69d58937 _pthread_mutex_firstfit_lock_slow + 222
29741615552516.590342 [1]  threadmon: 4   libddsc.0.7.0.dylib                 0x000000010e53d879 ddsrt_mutex_lock + 9
29751615552516.590343 [1]  threadmon: 5   libddsc.0.7.0.dylib                 0x000000010e4e8d51 lookup_proxy_topic + 33
29761615552516.590345 [1]  threadmon: 6   libddsc.0.7.0.dylib                 0x000000010e4dc59f handle_sedp + 335
29771615552516.590347 [1]  threadmon: 7   libddsc.0.7.0.dylib                 0x000000010e4dabc6 builtins_dqueue_handler + 2374
29781615552516.590348 [1]  threadmon: 8   libddsc.0.7.0.dylib                 0x000000010e4fb035 dqueue_thread + 405
29791615552516.590350 [1]  threadmon: 9   libddsc.0.7.0.dylib                 0x000000010e5061d3 create_thread_wrapper + 147
29801615552516.590351 [1]  threadmon: 10  libddsc.0.7.0.dylib                 0x000000010e53e0c9 os_startRoutineWrapper + 25
29811615552516.590353 [1]  threadmon: 11  libsystem_pthread.dylib             0x00007fff69d5d109 _pthread_start + 148
29821615552516.590354 [1]  threadmon: 12  libsystem_pthread.dylib             0x00007fff69d58b8b thread_start + 15
29831615552516.590356 [1]  threadmon: -- end of stack trace --
29841615552516.590358 [1]  threadmon: -- stack trace of dq.user requested --
29851615552516.591536 [1]  threadmon: -- stack trace follows --
29861615552516.591624 [1]  threadmon: 0   libddsc.0.7.0.dylib                 0x000000010e50effe log_stacktrace_sigh + 30
29871615552516.591633 [1]  threadmon: 1   libsystem_platform.dylib            0x00007fff69d515fd _sigtramp + 29
29881615552516.591635 [1]  threadmon: 2   ???                                 0x0000000000000000 0x0 + 0
29891615552516.591637 [1]  threadmon: 3   libddsc.0.7.0.dylib                 0x000000010e53d909 ddsrt_cond_wait + 9
29901615552516.591638 [1]  threadmon: 4   libddsc.0.7.0.dylib                 0x000000010e4faf98 dqueue_thread + 248
29911615552516.591640 [1]  threadmon: 5   libddsc.0.7.0.dylib                 0x000000010e5061d3 create_thread_wrapper + 147
29921615552516.591642 [1]  threadmon: 6   libddsc.0.7.0.dylib                 0x000000010e53e0c9 os_startRoutineWrapper + 25
29931615552516.591643 [1]  threadmon: 7   libsystem_pthread.dylib             0x00007fff69d5d109 _pthread_start + 148
29941615552516.591645 [1]  threadmon: 8   libsystem_pthread.dylib             0x00007fff69d58b8b thread_start + 15
29951615552516.591647 [1]  threadmon: -- end of stack trace --
29961615552516.591649 [1]  threadmon: -- stack trace of tev requested --
29971615552516.592742 [1]  threadmon: -- stack trace follows --
29981615552516.592791 [1]  threadmon: 0   libddsc.0.7.0.dylib                 0x000000010e50effe log_stacktrace_sigh + 30
29991615552516.592798 [1]  threadmon: 1   libsystem_platform.dylib            0x00007fff69d515fd _sigtramp + 29
30001615552516.592800 [1]  threadmon: 2   ???                                 0x00007fb8c1530d20 0x0 + 140431494155552
30011615552516.592801 [1]  threadmon: 3   libddsc.0.7.0.dylib                 0x000000010e53da2b ddsrt_cond_waitfor + 123
30021615552516.592803 [1]  threadmon: 4   libddsc.0.7.0.dylib                 0x000000010e50acce xevent_thread + 6190
30031615552516.592804 [1]  threadmon: 5   libddsc.0.7.0.dylib                 0x000000010e5061d3 create_thread_wrapper + 147
30041615552516.592806 [1]  threadmon: 6   libddsc.0.7.0.dylib                 0x000000010e53e0c9 os_startRoutineWrapper + 25
30051615552516.592807 [1]  threadmon: 7   libsystem_pthread.dylib             0x00007fff69d5d109 _pthread_start + 148
30061615552516.592809 [1]  threadmon: 8   libsystem_pthread.dylib             0x00007fff69d58b8b thread_start + 15
30071615552516.592811 [1]  threadmon: -- end of stack trace --
30081615552516.592813 [1]  threadmon: -- stack trace of recv requested --
30091615552516.594671 [1]  threadmon: -- stack trace follows --
30101615552516.594748 [1]  threadmon: 0   libddsc.0.7.0.dylib                 0x000000010e50effe log_stacktrace_sigh + 30
30111615552516.594753 [1]  threadmon: 1   libsystem_platform.dylib            0x00007fff69d515fd _sigtramp + 29
30121615552516.594755 [1]  threadmon: 2   ???                                 0x0000000000000000 0x0 + 0
30131615552516.594757 [1]  threadmon: 3   libddsc.0.7.0.dylib                 0x000000010e53d909 ddsrt_cond_wait + 9
30141615552516.594758 [1]  threadmon: 4   libddsc.0.7.0.dylib                 0x000000010e4fb4bb nn_dqueue_wait_until_empty_if_full + 75
30151615552516.594760 [1]  threadmon: 5   libddsc.0.7.0.dylib                 0x000000010e504f6d handle_regular + 1805
30161615552516.594761 [1]  threadmon: 6   libddsc.0.7.0.dylib                 0x000000010e501558 handle_submsg_sequence + 15224
30171615552516.594763 [1]  threadmon: 7   libddsc.0.7.0.dylib                 0x000000010e4fcfb2 do_packet + 706
30181615552516.594764 [1]  threadmon: 8   libddsc.0.7.0.dylib                 0x000000010e4fcc49 recv_thread + 2553
30191615552516.594765 [1]  threadmon: 9   libddsc.0.7.0.dylib                 0x000000010e5061d3 create_thread_wrapper + 147
30201615552516.594767 [1]  threadmon: 10  libddsc.0.7.0.dylib                 0x000000010e53e0c9 os_startRoutineWrapper + 25
30211615552516.594768 [1]  threadmon: 11  libsystem_pthread.dylib             0x00007fff69d5d109 _pthread_start + 148
30221615552516.594770 [1]  threadmon: 12  libsystem_pthread.dylib             0x00007fff69d58b8b thread_start + 15
30231615552516.594772 [1]  threadmon: -- end of stack trace --
30241615552516.594774 [1]  threadmon: -- stack trace of recvMC requested --
30251615552516.598504 [1]  threadmon: -- stack trace follows --
30261615552516.598564 [1]  threadmon: 0   libddsc.0.7.0.dylib                 0x000000010e50effe log_stacktrace_sigh + 30
30271615552516.598569 [1]  threadmon: 1   libsystem_platform.dylib            0x00007fff69d515fd _sigtramp + 29
30281615552516.598571 [1]  threadmon: 2   ???                                 0x0000000000000000 0x0 + 0
30291615552516.598572 [1]  threadmon: 3   libddsc.0.7.0.dylib                 0x000000010e4a6e11 ddsi_udp_conn_read + 193
30301615552516.598574 [1]  threadmon: 4   libddsc.0.7.0.dylib                 0x000000010e4fcdd6 do_packet + 230
30311615552516.598575 [1]  threadmon: 5   libddsc.0.7.0.dylib                 0x000000010e4fc3f9 recv_thread + 425
30321615552516.598577 [1]  threadmon: 6   libddsc.0.7.0.dylib                 0x000000010e5061d3 create_thread_wrapper + 147
30331615552516.598578 [1]  threadmon: 7   libddsc.0.7.0.dylib                 0x000000010e53e0c9 os_startRoutineWrapper + 25
30341615552516.598580 [1]  threadmon: 8   libsystem_pthread.dylib             0x00007fff69d5d109 _pthread_start + 148
30351615552516.598581 [1]  threadmon: 9   libsystem_pthread.dylib             0x00007fff69d58b8b thread_start + 15
30361615552516.598583 [1]  threadmon: -- end of stack trace --
30371615552516.598586 [1]  threadmon: -- stack trace of recvUC requested --
30381615552516.599773 [1]  threadmon: -- stack trace follows --
30391615552516.599831 [1]  threadmon: 0   libddsc.0.7.0.dylib                 0x000000010e50effe log_stacktrace_sigh + 30
30401615552516.599837 [1]  threadmon: 1   libsystem_platform.dylib            0x00007fff69d515fd _sigtramp + 29
30411615552516.599839 [1]  threadmon: 2   libddsc.0.7.0.dylib                 0x000000010e4bcfdd entidx_lookup_proxy_writer_guid + 61
30421615552516.599841 [1]  threadmon: 3   libddsc.0.7.0.dylib                 0x000000010e4a6e11 ddsi_udp_conn_read + 193
30431615552516.599842 [1]  threadmon: 4   libddsc.0.7.0.dylib                 0x000000010e4fcdd6 do_packet + 230
30441615552516.599844 [1]  threadmon: 5   libddsc.0.7.0.dylib                 0x000000010e4fc3f9 recv_thread + 425
30451615552516.599845 [1]  threadmon: 6   libddsc.0.7.0.dylib                 0x000000010e5061d3 create_thread_wrapper + 147
30461615552516.599847 [1]  threadmon: 7   libddsc.0.7.0.dylib                 0x000000010e53e0c9 os_startRoutineWrapper + 25
30471615552516.599849 [1]  threadmon: 8   libsystem_pthread.dylib             0x00007fff69d5d109 _pthread_start + 148
30481615552516.599894 [1]  threadmon: 9   libsystem_pthread.dylib             0x00007fff69d58b8b thread_start + 15
30491615552516.599899 [1]  threadmon: -- end of stack trace --
30501615552516.934485 [1]  threadmon: thread recv failed to make progress
30511615552522.403466 [1]  threadmon: thread recvUC failed to make progress
```

Which is caused by `tlm_register_with_proxy_endpoints_locked` locking the proxy participant while already holding `tl_admin_lock`. It turns out to be trivial to locally reproduce using:
```diff
diff --git a/src/core/ddsi/src/q_entity.c b/src/core/ddsi/src/q_entity.c
index 2b256cb8..96234e08 100644
--- a/src/core/ddsi/src/q_entity.c
+++ b/src/core/ddsi/src/q_entity.c
@@ -6130,6 +6130,7 @@ static void gc_delete_proxy_topic (struct gcreq *gcreq)
   struct gc_proxy_tp *gcdata = gcreq->arg;
 
   ddsrt_mutex_lock (&gcdata->proxypp->e.lock);
+  dds_sleepfor(DDS_MSECS(10));
   struct ddsi_domaingv *gv = gcdata->proxypp->e.gv;
   ddsrt_mutex_lock (&gv->topic_defs_lock);
   struct ddsi_topic_definition *tpd = gcdata->proxytp->definition;
```
I'll probably give it some more thought, but I wanted to record it here so I don't lose the information and so you could easily pick it up if you so desire.

eboasson commented:
An update: I don't actually see a reason for `tlm_register_with_proxy_endpoints_locked` to push the type definition into the proxy topic: it doesn't seem like that bit of information is ever used. I can understand why it wouldn't be used, I can also understand why general code hygiene would suggest it is better to update it anyway, but if that leads to a deadlock ...

dennis-adlink commented:
You're right, the sertype pointer in topic definitions was only used in the find remote topic implementation and can be worked around by a lookup of the sertype via the tl_admin using the type identifier stored in the topic definition (the overhead of this additional lookup is not a problem here).  Commit 993ed7d remove this field from `struct topic definition` to avoid the locking order issue when updating the definition of a proxy topic in case a new type is registered. 
I've rebased the PR branch on your branch https://github.com/eboasson/cyclonedds/tree/topic-discovery-pr615, so a force push was required. In the other two commits I've adjusted some time-outs and parameters in the topic discovery tests, so that these will pass also under high load. 

The failing test `ddsc_data_avail_stress_remote` on CI is unrelated to this PR I think, and probably caused by the discovery taking too long to connect the reader and writer and getting these in sync because of high loss configured in the test (33.3%) and other tests running in parallel. In another PR I'll update some parameters for this test to avoid this.

eboasson commented:
@dennis-adlink I think these latest changes indeed sort out the remaining issues, but now there is still some housekeeping to do before I can merge it: sorting out the conflicts and squashing the commits a bit (I don't see much value in keeping the various fixups as separate commits, but having them in that way is very convenient while still working on reviews).

Would you have time to do that in the near future?
